Invalid parsing of complex qualified path in expression

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I tried this code:

```rust
fn main() {
    let _ = <<F as Foo>::A as Bar>::f();
}
```

I expected to see this happen: no parsing error. `rustc` with `-Z parse-only` returns 0

Instead, this happened: 

```rust
test.rs:2:13: error: found unexpected token ‘<<’ in null denotation
    2 |     let _ = <<F as Foo>::A as Bar>::f();
      |             ^~
test.rs:2:15: error: failed to parse expression in let statement
    2 |     let _ = <<F as Foo>::A as Bar>::f();
      |               ^
test.rs:2:5: error: failed to parse statement or expression without block in block expression
    2 |     let _ = <<F as Foo>::A as Bar>::f();
      |     ^~~
test.rs:3:1: error: unrecognised token ‘}’ for start of item
    3 | }
      | ^
test.rs:3:1: error: failed to parse item in crate
```

or

```rust
test.rs:2:35: error: found null denotation scope resolution operator, and have not written handling for it
    2 |     let _ = <<F as Foo>::A as Bar>::f();
      |                                   ^~
test.rs:2:37: error: failed to parse expression in let statement
    2 |     let _ = <<F as Foo>::A as Bar>::f();
      |                                     ^
test.rs:2:5: error: failed to parse statement or expression without block in block expression
    2 |     let _ = <<F as Foo>::A as Bar>::f();
      |     ^~~
test.rs:3:1: error: unrecognised token ‘}’ for start of item
    3 | }
      | ^
test.rs:3:1: error: failed to parse item in crate
```

with basic handling of `<<` in `null_denotation`

### Meta

- What version of Rust GCC were you using: 1bd1e466453

--- a/gcc/rust/parse/rust-parse-impl.h
+++ b/gcc/rust/parse/rust-parse-impl.h
@@ -12650,6 +12650,13 @@ Parser<ManagedTokenSource>::parse_expr (int right_binding_power,
 	  || id == RIGHT_SQUARE)
 	return nullptr;
     }
+
+  if (current_token->get_id () == LEFT_SHIFT)
+    {
+      lexer.split_current_token (LEFT_ANGLE, LEFT_ANGLE);
+      current_token = lexer.peek_token ();
+    }
+
   lexer.skip_token ();
 
   // parse null denotation (unary part of expression)

--- a/gcc/testsuite/rust/compile/complex_qualified_path_in_expr.rs
+++ b/gcc/testsuite/rust/compile/complex_qualified_path_in_expr.rs
@@ -1,0 +1,34 @@
+struct S; // { dg-warning "struct is never constructed" "" { target *-*-* } .-1 }
+impl S {
+  fn f() -> i32 { return 0; }
+  // { dg-warning "associated function is never used" "" { target *-*-* } .-1 }
+}
+
+trait T1 {
+  type A;
+  fn f() -> i32 { return 0; }
+}
+
+impl T1 for S {
+  type A = S;
+}
+
+trait T2 {
+  type B;
+  fn f() -> i32 { return 0; }
+}
+
+impl T2 for S {
+  type B = S;
+}
+
+trait T3 {
+  fn f() -> i32 { return 0; }
+}
+
+impl T3 for S {}
+
+fn main() {
+  let _ = <<S as T1>::A as T2>::f();
+  let _ = <<<S as T1>::A as T2>::B as T3>::f();
+}

